main.tf

```diff
@@ -106,133 +231,342 @@ data "aws_iam_policy_document" "secrets_access" {
+      }
+      //      TODO test listener traffic
+      //      test_traffic_route {
+      //        listener_arns = aws_alb_listener.....
+      //      }
+      target_group {
+        name = aws_alb_target_group.blue.name
+      }
+      target_group {
+        name = aws_alb_target_group.green.name
+      }
+    }
+  }
+
+  lifecycle { ignore_changes = [blue_green_deployment_config] }
+}
+
+# ==================== CloudWatch ====================
+resource "aws_cloudwatch_log_group" "container_log_group" {
+  name              = local.cloudwatch_log_group_name
+  retention_in_days = var.log_retention_in_days
+  tags              = var.tags
+}
+
+# ==================== AutoScaling ====================
+resource "aws_appautoscaling_target" "main" {
   min_capacity       = var.min_capacity
   max_capacity       = var.max_capacity
-  resource_id        = "service/${module.fargate.ecs_cluster.name}/${module.fargate.ecs_service.name}"
+  resource_id        = "service/${aws_ecs_cluster.cluster.name}/${aws_ecs_service.service.name}"
   scalable_dimension = "ecs:service:DesiredCount"
   service_namespace  = "ecs"
-  step_up = {
+}
+resource "aws_appautoscaling_policy" "up" {
+  name               = "${var.app_name}-autoscale-up"
+  resource_id        = aws_appautoscaling_target.main.resource_id
+  scalable_dimension = aws_appautoscaling_target.main.scalable_dimension
+  service_namespace  = aws_appautoscaling_target.main.service_namespace
+
+  step_scaling_policy_configuration {
     adjustment_type         = "ChangeInCapacity"
+    metric_aggregation_type = "Average"
     cooldown                = 300
-    metric_aggregation_type = null // use default
-    step_adjustments = [
-      {
-        lower_bound        = 0
-        upper_bound        = null
-        scaling_adjustment = 1
-      }
-    ]
 
-    alarm = {
-      namespace = "AWS/ECS"
-      dimensions = {
-        ClusterName = module.fargate.ecs_cluster.name
-        ServiceName = module.fargate.ecs_service.name
-      }
-      statistic           = null // use default
-      metric_name         = "CPUUtilization"
-      comparison_operator = "GreaterThanThreshold"
-      threshold           = 75
-      period              = null // use default
-      evaluation_periods  = null // use default
+    step_adjustment {
+      scaling_adjustment          = 1
+      metric_interval_lower_bound = 0
     }
   }
+}
+resource "aws_cloudwatch_metric_alarm" "up" {
+  alarm_name = "${var.app_name}-alarm-up"
+  namespace  = "AWS/ECS"
+  dimensions = {
+    ClusterName = aws_ecs_cluster.cluster.name
+    ServiceName = aws_ecs_service.service.name
+  }
+  statistic           = "Average"
+  metric_name         = "CPUUtilization"
+  comparison_operator = "GreaterThanThreshold"
+  threshold           = 75
+  period              = 300
+  evaluation_periods  = 5
+  alarm_actions       = [aws_appautoscaling_policy.up.arn]
+  tags                = var.tags
+}
```

Comment:
I think this is fine for now, but we'll quickly need to allow people to customize this behavior (either customize this resource, or pass in their own resource).

Reply:
Implemented the ability to opt out of default auto scaling by defining `autoscaling_config` to null